### PR TITLE
Added Legend Option To Describe > Graph > BarChart dialogue 

### PR DESCRIPTION
**Pull request by derekagorhom**
fixes partly #8297 

I have added the legend option to the bar chart dialogue and also added the Reverse check box

@rdstern can you review this 

Thanks

### CODE REVIEW

**Comment by Vitalis95** [comment]
@derekagorhom , when I open the dialog for the first time I get the following error but it is already sorted in the PR #8395, once it is merged you can pull the changes

image

The checkbox and the input are not aligned

**Comment by Vitalis95** [comment]
@derekagorhom , let us have a quick chat we sort out some of the things before @rdstern  tests it because the graphs produced have changed from what was there before. I am available

**Comment by derekagorhom** [comment]
@rdstern can you review this.
Thanks

**Comment by rdstern** [comment]
@derekagorhom Sorry for the delay in reviewing this.  And now I would also like this to wait until the current changes are merged.  But I also suggest some changes to be ready.
a) Change Legend Position to just Legend.  
b) Make the Options box narrower and move it towards the Legends label.  (Still allow for the translations to be a longer word.)
c) Delete the Reverse option.  
d) Instead have another checkbox, default unchecked and Label Title:  Then (if checked) it adds the control for the legend title from the Titles sub-dialog and links to it.  (It is a shortcut for making the change there.)  

Thanks

**Review by rdstern — DISMISSED** [review]
@derekagorhom another I am approving.  @MeSophie this is also what you will be adding into the general graphics dialog.
@lloyddewit over to you.

**Comment by derekagorhom** [comment]
> @derekagorhom Sorry for the delay in reviewing this. And now I would also like this to wait until the current changes are merged. But I also suggest some changes to be ready. a) Change Legend Position to just Legend. b) Make the Options box narrower and move it towards the Legends label. (Still allow for the translations to be a longer word.) c) Delete the Reverse option. d) Instead have another checkbox, default unchecked and Label Title: Then (if checked) it adds the control for the legend title from the Titles sub-dialog and links to it. (It is a shortcut for making the change there.)
> 
> Thanks

@rdstern please are you still keen on having  these changes?

**Comment by rdstern** [comment]
@derekagorhom thanks for reminding me about this.  Could you make just the trivial change for now of just having the simpler Legend as the label and perhaps moving the pull down just a bit to the left, but leaving enough space in the label for the translation to be quite a lot wider than in English.

Leave the other suggestions for now.  
My reason is that the other changes are not essential.  As @lloyddewit is about to leave IDEMS I am keen to move as quickly as possible to a "finished" version of R-Instat.

**Comment by derekagorhom** [comment]
@rdstern I have made the requested changes, can you review them
Thanks

**Review by rdstern — DISMISSED** [review]
Looks fine

File: instat/dlgBarAndPieChart.vb [hunk 1]
```diff
@@ -77,8 +77,15 @@ Public Class dlgBarAndPieChart
     Private clsScaleSizeAreaFunction As New RFunction
     Private clsDummyFunction As New RFunction
     Private clsPointsFunction As New RFunction
+
+    Private clsGuideLegendFunction As New RFunction
+    Private clsGuideFunction As New RFunction
+    Private clsThemeFunction As New RFunction
+
+
```
Comment (Vitalis95) [inline instat/dlgBarAndPieChart.vb:85]: Remove

File: instat/dlgBarAndPieChart.vb [hunk 2]
```diff
@@ -77,8 +77,15 @@ Public Class dlgBarAndPieChart
...
     Private clsGeomLollipopAesFunction As New RFunction
 
+
```
Comment (Vitalis95) [inline instat/dlgBarAndPieChart.vb:88]: Remove

File: instat/dlgBarAndPieChart.vb [hunk 3]
```diff
@@ -132,9 +140,15 @@ Public Class dlgBarAndPieChart
         ucrPnlOptions.AddFunctionNamesCondition(rdoTreeMap, {"geom_treemap", "geom_treemap_text"})
         ucrPnlOptions.AddFunctionNamesCondition(rdoWordCloud, {"geom_text_wordcloud", "scale_size_area"})
 
+
```
Comment (Vitalis95) [inline instat/dlgBarAndPieChart.vb:143]: Remove

File: instat/dlgBarAndPieChart.vb [hunk 4]
```diff
@@ -132,9 +140,15 @@ Public Class dlgBarAndPieChart
...
+        ucrPnlOptions.AddToLinkedControls({ucrReceiverX, ucrChkReorderValue, ucrChkLollipop}, {rdoValue}, bNewLinkedAddRemoveParameter:=True, bNewLinkedHideIfParameterMissing:=True)
+        ucrPnlOptions.AddToLinkedControls(ucrChkReorderFrequency, {rdoFrequency}, bNewLinkedHideIfParameterMissing:=True)
+
         ucrPnlOptions.AddToLinkedControls({ucrChkFlipCoordinates, ucrChkPolarCoordinates, ucrReceiverByFactor, ucrInputBarChartPositions, ucrChkAddLabelsText, ucrVariablesAsFactorForBarChart, ucrChkBacktoback}, {rdoFrequency, rdoValue}, bNewLinkedAddRemoveParameter:=True, bNewLinkedHideIfParameterMissing:=True)
         ucrPnlOptions.AddToLinkedControls({ucrReceiverX, ucrChkReorderValue, ucrInputAddReorder, ucrChkLollipop}, {rdoValue}, bNewLinkedAddRemoveParameter:=True, bNewLinkedHideIfParameterMissing:=True)
         ucrPnlOptions.AddToLinkedControls({ucrChkReorderFrequency, ucrInputAddReorder}, {rdoFrequency}, bNewLinkedHideIfParameterMissing:=True)
+
```
Comment (Vitalis95) [inline instat/dlgBarAndPieChart.vb:151]: remove

File: instat/dlgBarAndPieChart.vb [hunk 5]
```diff
@@ -295,6 +309,23 @@ Public Class dlgBarAndPieChart
...
+        dctLegendPosition.Add("None", Chr(34) & "none" & Chr(34))
+        dctLegendPosition.Add("Left", Chr(34) & "left" & Chr(34))
+        dctLegendPosition.Add("Right", Chr(34) & "right" & Chr(34))
+        dctLegendPosition.Add("Top", Chr(34) & "top" & Chr(34))
+        dctLegendPosition.Add("Bottom", Chr(34) & "bottom" & Chr(34))
+        ucrInputLegendPosition.SetItems(dctLegendPosition)
+        ucrChkLegendPosition.AddParameterPresentCondition(True, "legend.position")
+        ucrChkLegendPosition.AddParameterPresentCondition(False, "legend.position", False)
+
+
```
Comment (Vitalis95) [inline instat/dlgBarAndPieChart.vb:328]: remove

File: instat/dlgBarAndPieChart.vb [hunk 6]
```diff
@@ -407,8 +438,15 @@ Public Class dlgBarAndPieChart
         clsScaleSizeAreaFunction = New RFunction
         clsDummyFunction = New RFunction
         clsPointsFunction = New RFunction
+
+        clsGuideLegendFunction = New RFunction
+        clsGuideFunction = New RFunction
+        clsThemeFunction = GgplotDefaults.clsDefaultThemeFunction
+
+
```
Comment (Vitalis95) [inline instat/dlgBarAndPieChart.vb:446]: remove

File: instat/dlgBarAndPieChart.vb [hunk 7]
```diff
@@ -407,8 +438,15 @@ Public Class dlgBarAndPieChart
...
         clsGeomLollipopAesFunction = New RFunction
 
+
```
Comment (Vitalis95) [inline instat/dlgBarAndPieChart.vb:449]: remove

File: instat/dlgBarAndPieChart.vb [hunk 8]
```diff
@@ -431,10 +469,16 @@ Public Class dlgBarAndPieChart
 
         clsDummyFunction.AddParameter("Checked", "FALSE", iPosition:=0)
 
+        clsGuideLegendFunction.SetRCommand("guide_legend")
+        clsGuideLegendFunction.AddParameter("reverse", "TRUE")
```
Comment (Vitalis95) [inline instat/dlgBarAndPieChart.vb:473]: `iposition=0` ?

File: instat/dlgBarAndPieChart.vb [hunk 9]
```diff
@@ -431,10 +469,16 @@ Public Class dlgBarAndPieChart
...
+
         clsBaseOperator.SetOperation("+")
         clsBaseOperator.AddParameter("ggplot", clsRFunctionParameter:=clsRggplotFunction, iPosition:=0)
         clsBaseOperator.AddParameter("geom_bar", clsRFunctionParameter:=clsRgeomBarFunction, iPosition:=2)
 
+        clsGuideFunction.SetRCommand("guides")
+        clsGuideFunction.AddParameter("fill", clsRFunctionParameter:=clsGuideLegendFunction)
```
Comment (Vitalis95) [inline instat/dlgBarAndPieChart.vb:480]: `iposition=0`

File: instat/dlgBarAndPieChart.vb [hunk 10]
```diff
@@ -631,11 +675,17 @@ Public Class dlgBarAndPieChart
         ucrNudMaxSize.SetRCode(clsScaleSizeAreaFunction, bReset)
         ucrChkReorderFrequency.SetRCode(clsDummyFunction, bReset)
 
+        ucrChkReverse.SetRCode(clsBaseOperator, bReset)
+        ucrChkLegendPosition.SetRCode(clsThemeFunction, bReset, bCloneIfNeeded:=True)
+        ucrInputLegendPosition.SetRCode(clsThemeFunction, bReset, bCloneIfNeeded:=True)
+
+
```
Comment (Vitalis95) [inline instat/dlgBarAndPieChart.vb:682]: remove

File: instat/dlgBarAndPieChart.vb [hunk 11]
```diff
@@ -631,11 +675,17 @@ Public Class dlgBarAndPieChart
...
         If bReset Then
             ucrChkStart.SetRCode(clsGeomTreemapFunction, bReset)
             ucrChkLayout.SetRCode(clsGeomTreemapFunction, bReset)
             ucrChkIncreaseSize.SetRCode(clsScaleSizeAreaFunction, bReset)
         End If
+
```
Comment (Vitalis95) [inline instat/dlgBarAndPieChart.vb:688]: remove

File: instat/dlgBarAndPieChart.vb [hunk 12]
```diff
@@ -762,6 +812,11 @@ Public Class dlgBarAndPieChart
             ucrVariablesAsFactorForBarChart.RemoveIncludedMetadataProperty("class")
             ucrVariablesAsFactorForBarChart.strSelectorHeading = "Variables"
             ucrInputBarChartPositions.Visible = Not ucrReceiverByFactor.IsEmpty()
+
+            ucrChkReverse.Visible = Not ucrReceiverByFactor.IsEmpty()
+
```
Comment (Vitalis95) [inline instat/dlgBarAndPieChart.vb:817]: remove all empty lines here

File: instat/dlgBarAndPieChart.vb [hunk 13]
```diff
@@ -1159,4 +1214,27 @@ Public Class dlgBarAndPieChart
             clsBaseOperator.RemoveParameterByName("geom_treemap_text")
         End If
     End Sub
+
+
```
Comment (Vitalis95) [inline instat/dlgBarAndPieChart.vb:1218]: remove

File: instat/dlgBarAndPieChart.vb [hunk 14]
```diff
@@ -1159,4 +1214,27 @@ Public Class dlgBarAndPieChart
...
+            clsBaseOperator.AddParameter("theme", clsRFunctionParameter:=clsThemeFunction, iPosition:=15)
+        Else
+            clsBaseOperator.RemoveParameterByName("theme")
+        End If
+    End Sub
+
+    Private Sub ucrChkLegendPosition_ControlValueChanged(ucrChangedControl As ucrCore) Handles ucrChkLegendPosition.ControlValueChanged
+        AddRemoveTheme()
+    End Sub
+
```
Comment (Vitalis95) [inline instat/dlgBarAndPieChart.vb:1238]: remove both empty lines

**Comment by rdstern** [comment]
@Fidel365 I am afraid I have a couple of details and then a slightly larger suggestion, which may need help from @N-thony;

a) Could you delete the Colon in `Legend: `.  I am sorry I suggested that, because I now realise it is not consistent with the other labels on the dialog.  
b) I can't see, on the dialog, how wide the space is for the legend label.  Make it right up to the start of the drop down, so there is enough space if the translated word is longer in other languages.  Perhaps even move the pull-down right just a little.
c) I have been a slow-learner on the Add Reverse.  But now I understand it I am very happy you have persisted!  Please again change Add Reverse to just say Reverse.
d) That's all cosmetic, and N-Thony may decide we can merge then.  But we now have the Legend control in 2 places and this has introduced new problems and opportunities:
1) The general place is in Plot Options > Themes.  There it currently says Legend Position.  Let's be consistent, and change that to Legend.  
2) Then move the drop down left a bit, and add the Reverse option there as well.  (We can have less on the main dialog, than the sub-dialog, but this is now the opposite.)
3) I think that when an option is available in both places, then they are synchronised?  @N-thony am I right on this point?  So, if I make a change to the Legend position and setting of reverse on the main dialog, then it will be reflected also on the sub-dialog - and vv?

**Comment by derekagorhom** [comment]
@rdstern, I have made the cosmetic changes on the Bar chart and also done some code changes on the sub-dialog.
can you review this again. 
Thanks

**Comment by rdstern** [comment]
@derekagorhom it is possible we should abandon the extra of the legend on the bar chart.  It is taking too long.
But it must work well on the Themes sub-dialog:
image

Please the reverse should not be visible when the legend is not ticked.

image

Then None appears to work, but every other option gives the legend on the right.
Reverse appears to work and that's good.

I am now getting concerned that making the changes to the sub-dialog - which we have to get right - and also making a simple way of getting the options on the bar chart main dialog, and synchronising them, is taking too long.

So my suggestion is that you abandon the pull request, so there is no legend option on the main bar chart dialog for now.  Go back to the original version of the sub-dialog and add thge changed, so the reverse is included and the label shortened.  Then let's get this merged.  There is too much other work to keep spending time on this degtailed point.  Users will have to accept the sub-dialog for these legend options.

**Comment by derekagorhom** [comment]
@rdstern , @N-thony I am closing this PR

**Pull request closed without merging**